common/src/chain/transaction/signature/tests/sign_and_mutate.rs

```diff
@@ -0,0 +1,509 @@
+        remove_last_input,
+        remove_first_output,
+    ];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        for input in 0..inputs - 1 {
+            assert_eq!(
+                verify_signature(&destination, &tx, input),
+                Err(TransactionSigError::SignatureVerificationFailed)
+            );
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs)),
+        );
+    }
+
+    let mutations = [add_output, remove_last_output];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        for input in 0..inputs - 1 {
+            assert_eq!(verify_signature(&destination, &tx, input), Ok(()));
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs)),
+        );
+    }
+
+    {
+        let tx = modify_output(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        assert_eq!(
+            verify_signature(&destination, &tx, 0),
+            Err(TransactionSigError::SignatureVerificationFailed),
+        );
+        for input in 1..inputs - 1 {
+            assert_eq!(verify_signature(&destination, &tx, input), Ok(()),);
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs)),
+        );
+    }
+}
+
+#[test]
+fn mutate_single_anyonecanpay() {
+    let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+    let destination = Destination::PublicKey(public_key);
+    let sighash_type =
+        SigHashType::try_from(SigHashType::SINGLE | SigHashType::ANYONECANPAY).unwrap();
+    let tx =
+        generate_and_sigh_tx(&destination, INPUTS, OUTPUTS, &private_key, sighash_type).unwrap();
+
+    let mutations = [add_input, remove_last_input, add_output, remove_last_output];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        for input in 0..inputs - 1 {
+            assert_eq!(
+                verify_signature(&destination, &tx, input),
+                Ok(()),
+                "{input}"
+            );
+        }
+        assert_eq!(
+            verify_signature(&destination, &tx, inputs),
+            Err(TransactionSigError::InvalidInputIndex(inputs, inputs))
+        );
+    }
+
+    let mutations = [modify_input, modify_output];
+    for mutate in mutations.into_iter() {
+        let tx = mutate(&tx);
+        let inputs = tx.inputs().unwrap().len();
+
+        assert_eq!(
+            verify_signature(&destination, &tx, 0),
+            Err(TransactionSigError::SignatureVerificationFailed),
+        );
+        for input in 1..inputs - 1 {
```

Comment:
These random ranges are basically playing with fire... we should use something more imperative than something that can be wrong by coincidence.

Reply:
I agree that is somewhat controversial, but I still think it is better than checking for arbitrary inputs as before.